plugins/workload-management/src/main/java/org/opensearch/plugin/wlm/service/QueryGroupPersistenceService.java

```diff
@@ -0,0 +1,194 @@
+    ) {
+        this.clusterService = clusterService;
+        this.createQueryGroupThrottlingKey = clusterService.registerClusterManagerTask(CREATE_QUERY_GROUP_THROTTLING_KEY, true);
+        setMaxQueryGroupCount(MAX_QUERY_GROUP_COUNT.get(settings));
+        clusterSettings.addSettingsUpdateConsumer(MAX_QUERY_GROUP_COUNT, this::setMaxQueryGroupCount);
+    }
+
+    /**
+     * Set maxQueryGroupCount to be newMaxQueryGroupCount
+     * @param newMaxQueryGroupCount - the max number of QueryGroup allowed
+     */
+    public void setMaxQueryGroupCount(int newMaxQueryGroupCount) {
+        validateMaxQueryGroupCount(newMaxQueryGroupCount);
+        this.maxQueryGroupCount = newMaxQueryGroupCount;
+    }
+
+    /**
+     * Validator for maxQueryGroupCount
+     * @param maxQueryGroupCount - the maxQueryGroupCount number to be verified
+     */
+    private static void validateMaxQueryGroupCount(int maxQueryGroupCount) {
+        if (maxQueryGroupCount > DEFAULT_MAX_QUERY_GROUP_COUNT_VALUE || maxQueryGroupCount < MIN_QUERY_GROUP_COUNT_VALUE) {
+            throw new IllegalArgumentException(QUERY_GROUP_COUNT_SETTING_NAME + " should be in range [1-100].");
+        }
+    }
+
+    /**
+     * Update cluster state to include the new QueryGroup
+     * @param queryGroup {@link QueryGroup} - the QueryGroup we're currently creating
+     * @param listener - ActionListener for CreateQueryGroupResponse
+     */
+    public void persistInClusterStateMetadata(QueryGroup queryGroup, ActionListener<CreateQueryGroupResponse> listener) {
+        clusterService.submitStateUpdateTask(SOURCE, new ClusterStateUpdateTask(Priority.NORMAL) {
+            @Override
+            public ClusterState execute(ClusterState currentState) throws Exception {
+                return saveQueryGroupInClusterState(queryGroup, currentState);
+            }
+
+            @Override
+            public ThrottlingKey getClusterManagerThrottlingKey() {
+                return createQueryGroupThrottlingKey;
+            }
+
+            @Override
+            public void onFailure(String source, Exception e) {
+                logger.warn("failed to save QueryGroup object due to error: {}, for source: {}.", e.getMessage(), source);
+                listener.onFailure(e);
+            }
+
+            @Override
+            public void clusterStateProcessed(String source, ClusterState oldState, ClusterState newState) {
+                CreateQueryGroupResponse response = new CreateQueryGroupResponse(queryGroup, RestStatus.OK);
+                listener.onResponse(response);
+            }
+        });
+    }
+
+    /**
+     * This method will be executed before we submit the new cluster state
+     * @param queryGroup - the QueryGroup we're currently creating
+     * @param currentClusterState - the cluster state before the update
+     */
+    public ClusterState saveQueryGroupInClusterState(final QueryGroup queryGroup, final ClusterState currentClusterState) {
```

Comment:
This can be package private. Assuming it is not private for testing purpose?

Reply:
It is even possible to write tests using latch and `ClusterStateListener` to not have this even package private. But, we can skip that for now